@rules as coroutines

stuhood opened the pull request:
### Problem

The "complex" `Selector` types used to select `@rule` inputs for subjects other than the current subject of an `@rule` have proven to be entirely too complicated. Additionally, their lack of flexibility meant that we needed to introduce many intermediate types to provide links between `@rules` in cases where imperative logic was needed for filtering or conditionals.

The only remaining motivating factor for these complex selectors was that they allow for closed-world compilation of the `@rule` graph, which avoids the need for error checking in rule bodies and provides for easier composition of sets of `@rules`.

### Solution

After experimenting with using coroutines (ie: functions containing `yield` statements that receive a result on their left hand side), it became clear both that:

1. Coroutines were universally easier to understand than the Selector DSL
2. The dependency requests made by coroutines could still easily be statically checked via a small amount of runtime introspection.

And so, this change removes all usage of `SelectProjection` and most usage of `SelectDependencies` in favor of `@rule` coroutines. It does not yet remove the implementation of those `Selectors` (although doing so will shed a massive amount of complexity from the codebase).

### Result

`@rules` are still statically checked, but are now much easier to write. Performance is unaffected.

cosmicexplorer commented on src/python/pants/engine/build_files.py:
```diff
@@ -114,17 +96,16 @@ def _raise_did_you_mean(address_family, name):
   dependencies field, since those are requested explicitly by tasks using SelectDependencies.
   """
 
+  address_family = yield Get(AddressFamily, Dir(address.spec_path))
```
If no `Dir -> AddressFamily` path exists in the rule graph, does it error at this line? Would the previous use of `SelectProjection` have errored out earlier, during the rule graph construction?

stuhood replied:
Yep, exactly. And that's the primary downside of this change as it stands: we would lose the validation that all rules are satisfiable.

Have been thinking more about this though, and I think that with some very, very basic parsing (essentially, "capture the RHS of yield statements"), we could get back the static checking... which might be justified.

stuhood replied:
...and also, as it stands the exception would be created in the rust code rather than literally on this line. It looks like it is easy to send an exception to the generator though with `generator.throw()`.

cosmicexplorer replied:
I think "capture the RHS of yield statements" is a highly reasonable goal, but I don't know if it needs to be done by parsing -- that being said I remember python's ast module is supposed to be not awful? I can't see what the exact code would look like right now, but since we know we _can_ get the desired info we want before invoking the python rule definition (with the previous example using `SelectProjection`), I think there's hopefully a middle ground where we can specify rule graphs without too much ceremony (probably exactly the same in spirit as parsing for yield statements). I would love to parse some python for yield statements if it gets us there, but I think there might be a neater way -- will think on it.

stuhood replied:
This is now  resolved: all `Get` statements in the body of an `@rule` are statically checked at `RuleGraph` construction time.

stuhood commented:
Parsing `yield` statements from `@rule`s appears to be straightforward: have added one more commit on top that demonstrates it in a single file.

So I think that with one relatively minor restriction (which technically already existed) we can preserve the static checks; namely: the request/Get/yield would need to include both the product and subject type explicitly. An example syntax might be:
```
# verbose form: Get(product_type, subject_type, subject)
classpath = yield Get(Classpath, ScalaSources, sources)
# potential shorthand: Get(product_type, subject_type(subject))
classpath = yield Get(Classpath, ScalaSources(snapshot))
```

illicitonion commented:
I like it, much cleaner than `SelectProjection` imo :)

stuhood commented:
I've implemented the static checks for `Get` calls, so this is almost ready to graduate from experiment to recommend API.

The next order of business: bikeshedding. What should `Get`'s final name be?

illicitonion commented on src/rust/engine/src/nodes.rs:
```diff
@@ -356,7 +372,7 @@ impl Select {
             variants: self.variants.clone(),
             task: task,
-            entry: entry.clone(),
+            entry: Arc::new(entry.clone()),
```
Seems a little strange for entry to be both `Arc`'d _and_ `clone`'d here. My instinct (which may be wrong) is that either it's a cheap to clone type, and we shouldn't be `Arc`ing it, or it's an expensive to clone type, and we should always be `Arc`ing it...

stuhood replied:
Yea, I have a separate review that attempts to Arc up the rule-graph, since it is full of tiny `Vec`s. On the other hand, it does not appear to impact performance in a significant way... so perhaps this would not either. It's very surprising to me that all of the allocation doesn't have a noticeable impact.

illicitonion commented:
Very cool :) Though I'm not sure I entirely follow the fine details of everything going on.

I'm happy to keep the name `Get` - we can always change it in the future if we need to (I'm assuming this will be an "unstable" API for a bit :))

My next request would be a few Python unit tests which should this actually in use

stuhood commented:
I haven't written any new unit tests, but I updated the documentation in the topmost commit, and added validation to the `Get` constructor, so I believe that this is now reviewable.

stuhood commented:
A link to the rendered README: https://github.com/twitter/pants/blob/56dd7bdd4df03cc8e6a7864bf1ab8231a2500656/src/python/pants/engine/README.md

jsirois commented:
I like the direction, but I'm not in love with the new syntax / concept / mixed-point dependency injection. Perhaps best to explain with a summary example. From the current change we now have:
```python
class Get(datatype('Get', ['product_type', 'subject'])):
  pass

@rule(UnhydratedStruct, [Select(AddressMapper), Select(Address)])
def resolve_unhydrated_struct(address_mapper, address):
  address_family = yield Get(AddressFamily, Dir(address.spec_path)
  ...
```

So we have dependency injection through function arguments and, now, co-routine results.

This could alternatively be modeled with just dependency injection through function aruments if we allow for both terminal values _and_ rules (functins) to be injected:
```python
class Rule(datatype('Rule', ['product_type', 'subject_type'])):
  def __call__(subject):
    pass  # IE: Get
  
  def map(subjects):
    pass  # IE: GetMulti

@rule(UnhydratedStruct, [Select(AddressMapper), Select(Address), Rule(AddressFamily, Dir)])
def resolve_unhydrated_struct(address_mapper, address, address_family_rule):
  address_family = address_family_rule(address.spec_path)
  ...
```

IOW, I'm a Rule author who know how to operate on certain selected values but I need to enlist the help of other rules imperatively. The proposed syntax change says this more directly I think and allows eliminating AST gymnastics.

stuhood commented:
Thanks @jsirois , that's definitely an interesting suggestion!

One downside I see is that it removes the trampolining that happens when we `yield` as a coroutine. `@rules` actually synchronously calling one another would put us back in a python threadstack ~all of the time. Whether this results in a net advantage in terms of time spent outside of the GIL is an open question, because if python functions were to synchronously call into other rules, I believe that the GIL would be released for the length of the call.

> Taken further (but importantly ignoring variant selects), this just means @rule becomes nothing but type annotations for a python function; ie: things eventuially could just boil down to:

I believe that it's possible that the `@rule` declaration can actually be simplified with either of the proposals, but I haven't looked into how variants will affect it either. I'd need to look at more callsites, but it's possible that `SelectVariant` would become a flavor of `Get`, which would mean that `@rule` arguments could be simplified similarly.

stuhood commented:
Another advantage that @illicitonion might point out is that avoiding coroutines would mean that we could more easily use Skylark at some point (which doesn't support them, afaik).

jsirois commented:
I may be missing somethin here, but couldn't the implementation of the call into the engine be masked, deferring the optimizations for behind-the-scenes work?:
```python
>>> def coroutine(value):
...   result = yield rust(value)
...   yield result
... 
>>> def rust(value):
...   return value * 2
... 
>>> def using_coroutine(value):
...   c = coroutine(value)
...   for r in c:
...     if r:
...       return r
... 
>>> using_coroutine(3)
6
```

stuhood commented:
> I may be missing something here, but couldn't the implementation of the call into the engine be masked, deferring the optimizations for behind-the-scenes work?:

`result = yield rust(value)` is still a "python calls rust" situation... so the coroutine would not actually be doing in anything useful with the `yield`. In this PR the `@rule` yields to the rust code which sends back the value. Again, really not certain that it buys us less time under the GIL. But it's recursion safe at least?

stuhood commented:
After dancing around the realization a bit, it occurred to us this morning that if we have synchronous calls back and forth between python and rust, we can't have async io. While fully switching to synchronous io via `rayon` would be a possibility, it would be a pretty massive refactor... and there is a lot of good stuff coming down the pipe for async in rust.

stuhood commented:
The top commit now adds a test utility function that makes it easier to test `@rules` that make `Get` requests.
